### PR TITLE
[MP] Add L0<->L1 throughput metrics (store/load GB/s)

### PR DESCRIPTION
**Pull request by Oasis-Git**
Adds two OTel histograms exposed via Prometheus:

- lmcache_mp_l0_l1_store_throughput_gbs  — GPU->CPU (L0->L1)
- lmcache_mp_l0_l1_load_throughput_gbs   — CPU->GPU (L1->L0)

Both carry `engine_id` and `gpu_id` attributes so operators can distinguish per-vLLM-worker throughput within a shared MP server.

Implementation:

- server.py: augment MP_{STORE,RETRIEVE}_{START,END} event metadata with engine_id (= instance_id), gpu_id (= device.index), and total_bytes on END events (summed from reserved_dict / memory_objs). START/END fire via publish_on_stream, so timestamps reflect true GPU-stream copy time.
- L0L1ThroughputSubscriber correlates START->END pairs by session_id, samples per-request at START (default 1%), records GB/s on END.
- Registered in init_observability under metrics_enabled.
- Unit tests cover subscription surface, happy path, sampling behavior, and edge cases (orphan END, zero bytes, zero duration, etc.).

**What this PR does / why we need it**:

**Special notes for your reviewers**:

**If applicable**:

- [ ] this PR contains user facing changes - docs added
- [ ] this PR contains unit tests

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it changes MP `store`/`retrieve` event metadata contracts and adds extra bookkeeping (`total_bytes`, `engine_id`, `model_name`) on the hot GPU-copy path, which could impact downstream subscribers if assumptions differ.
> 
> **Overview**
> Adds **sampled per-request GPU↔CPU throughput metrics** by introducing `L0L1ThroughputSubscriber`, which correlates `MP_{STORE,RETRIEVE}_{START,END}` pairs and records two new OTel histograms: `lmcache_mp.l0_l1_store_throughput_gbs` and `lmcache_mp.l0_l1_load_throughput_gbs`.
> 
> Updates MP server `store`/`retrieve` to enrich START/END events with `engine_id`/`model_name` and to include `total_bytes` on END events so throughput can be computed accurately from GPU-stream timestamps. Registers the new subscriber in observability config, documents the new event/metric contracts, and adds a comprehensive unit test suite covering sampling and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9cccfc626e743979dce094befba7c53abcef7224. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: lmcache/v1/multiprocess/server.py [hunk 1]
```diff
@@ -319,7 +319,11 @@ def store(
                 Event(
                     event_type=EventType.MP_STORE_START,
                     session_id=key.request_id,
-                    metadata={"device": str(gpu_context.device)},
+                    metadata={
+                        "device": str(gpu_context.device),
+                        "engine_id": instance_id,
+                        "gpu_id": gpu_context.device.index,
```
Comment (ApostaC) [inline lmcache/v1/multiprocess/server.py]: Since we already have `device` as the attribute, we don't need to put `gpu_id` here anymore.

File: lmcache/v1/multiprocess/server.py [hunk 2]
```diff
@@ -373,6 +377,7 @@ def store(
                         self.storage_manager.finish_write,
                         list(reserved_dict.keys()),
                     )
+                total_bytes = sum(mo.get_size() for mo in reserved_dict.values())
```
Comment (ApostaC) [inline lmcache/v1/multiprocess/server.py]: Potential optimization: we can run get_size for a single object and times it by the number of objects

File: lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        assert 0 < sample_rate <= 1.0, (
+            f"sample_rate must be in (0, 1.0], got {sample_rate}"
+        )
+        self._sample_rate = sample_rate
+
+        # session_id -> t_start. Populated only for sampled sessions.
+        self._pending_store: dict[str, float] = {}
+        self._pending_load: dict[str, float] = {}
+
+        meter = metrics.get_meter("lmcache.l0_l1")
```
Comment (ApostaC) [inline lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py]: Maybe let's call it `lmcache_mp.perf`? Other perf-related metrics can reuse the same meter name maybe. WDYT?

File: lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        if dt <= 0:
+            return
+
+        engine_id = event.metadata.get("engine_id")
+        gpu_id = event.metadata.get("gpu_id")
+        attrs: dict[str, Any] = {}
+        if engine_id is not None:
+            attrs["engine_id"] = str(engine_id)
+        if gpu_id is not None:
+            attrs["gpu_id"] = str(gpu_id)
```
Comment (ApostaC) [inline lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py]: Attr needs to include `lmcache_instance_id` and `model_name`?

File: lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+            EventType.MP_RETRIEVE_END: self._on_retrieve_end,
+        }
+
+    # -- Store path (L0→L1, GPU→CPU) ---------------------------------------
+
+    def _on_store_start(self, event: Event) -> None:
+        if random.random() >= self._sample_rate:
+            return
+        if event.session_id:
+            self._pending_store[event.session_id] = event.timestamp
```
Comment (ApostaC) [inline lmcache/v1/mp_observability/subscribers/metrics/l0_l1_throughput.py]: Potential problem: duplicated keys -- need to use (event.session_id, gpu_id)

**Review by ApostaC — COMMENTED** [review]
LGTM! Waiting for instance id pr

**Review by ApostaC — APPROVED** [review]
LGTM!

**Review by royyhuang — APPROVED** [review]
LGTM
